java-checks/src/main/java/org/sonar/java/checks/SillyEqualsCheck.java

```diff
@@ -0,0 +1,120 @@
+  priority = Priority.MAJOR)
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.LOGIC_RELIABILITY)
+@SqaleConstantRemediation("15min")
+public class SillyEqualsCheck extends AbstractMethodDetection {
+
+  private static final String JAVA_LANG_OBJECT = "java.lang.Object";
+
+  private static final String MESSAGE = "Remove this call to \"equals\"; comparisons between unrelated types always return false.";
+
+  @Override
+  protected List<MethodInvocationMatcher> getMethodInvocationMatchers() {
+    return ImmutableList.of(MethodInvocationMatcher.create()
+      .name("equals")
+      .addParameter(JAVA_LANG_OBJECT));
+  }
+
+  @Override
+  protected void onMethodFound(MethodInvocationTree tree) {
+    ExpressionTree firstArgument = Iterables.getOnlyElement(tree.arguments());
+    Type argumentType = firstArgument.symbolType().erasure();
+    Type ownerType = getMethodOwnerType(tree).erasure();
+    if (isLiteralNull(firstArgument)) {
+      addIssue(tree, "Remove this call to \"equals\"; comparisons against null always return false; consider using '== null' to check for nullity.");
+    } else if (ownerType.isArray()) {
+      checkWhenOwnerIsArray(tree, (Type.ArrayType) ownerType, argumentType);
+    } else {
+      checkWhenOwnerIsNotArray(tree, ownerType, argumentType);
+    }
+  }
+
+  private void checkWhenOwnerIsArray(Tree tree, Type.ArrayType ownerType, Type argumentType) {
+    if (argumentType.isArray()) {
+      if (areNotRelated(ownerType.elementType(), ((Type.ArrayType) argumentType).elementType())) {
+        addIssue(tree, "Remove this call to \"equals\"; comparisons between unrelated arrays always return false.");
+      } else {
+        addIssue(tree, "Use \"Arrays.equals(array1, array2)\" or the \"==\" operator instead of using the \"Object.equals(Object obj)\" method.");
+      }
+    } else if (!argumentType.is(JAVA_LANG_OBJECT)) {
+      addIssue(tree, "Remove this call to \"equals\"; comparisons between an array and a type always return false.");
+    }
+  }
+
+  private void checkWhenOwnerIsNotArray(Tree tree, Type ownerType, Type argumentType) {
+    if (argumentType.isArray() && !ownerType.is(JAVA_LANG_OBJECT)) {
+      addIssue(tree, "Remove this call to \"equals\"; comparisons between a type and an array always return false.");
+    } else if (argumentType.isClass() && areNotRelated(ownerType, argumentType)
+      && (areTypesFinalClassAndInterface(ownerType, argumentType) || areNeitherInterfaces(ownerType, argumentType))) {
+      addIssue(tree, MESSAGE);
+    }
+  }
+
+  private boolean areNeitherInterfaces(Type ownerType, Type argumentType) {
+    return !ownerType.symbol().isInterface() && !argumentType.symbol().isInterface();
+  }
+
+  private boolean areTypesFinalClassAndInterface(Type ownerType, Type argumentType) {
+    return (ownerType.symbol().isInterface() && argumentType.symbol().isFinal()) || (argumentType.symbol().isInterface() && ownerType.symbol().isFinal());
```

Comment:
Shouldn't this be a call to isClass instead of isInterface ? And if yes, it means there is a lack of tests.